Zend/zend_objects.c

```diff
@@ -287,9 +287,45 @@ ZEND_API void ZEND_FASTCALL zend_objects_clone_members(zend_object *new_object,
 		} ZEND_HASH_FOREACH_END();
 	}
 
-	if (has_clone_method) {
+	if (might_update_properties) {
 		GC_ADDREF(new_object);
-		zend_call_known_instance_method_with_0_params(new_object->ce->clone, new_object, NULL);
+		if (old_object->ce->clone) {
+			zend_call_known_instance_method_with_0_params(new_object->ce->clone, new_object, NULL);
+		}
+
+		if (EXPECTED(!EG(exception)) && zend_hash_num_elements(properties) > 0) {
+			/* Unlock readonly properties once more. */
+			if (ZEND_CLASS_HAS_READONLY_PROPS(new_object->ce) && old_object->ce->clone) {
```

Comment:
Do we need to unlock all properties again? After all, multiple writes to to readonly properties are forbidden within `__clone`. This problem can still occur with something like:

```php
class C {
    public function __construct(
        private readonly string $prop,
    ) {}
    public string $a { set { $this->prop = $value; } }
    public string $b { set { $this->prop = $value; } }
}

var_dump(clone(new C('foo'), [
    'a' => 'a',
    'b' => 'b',
]));
```

Reply:
Regardless, this can be skipped if `!old_object->ce->clone` as nothing happened between unlocking and this path.

Reply:
> Do we need to unlock all properties again?

I found it easiest to just unlock them again, instead of trying to find the property fields corresponding to the “with” properties.

> Regardless, this can be skipped if !old_object->ce->clone as nothing happened between unlocking and this path.

This condition is already there, or did I misunderstand?

Reply:
> I found it easiest to just unlock them again, instead of trying to find the property fields corresponding to the “with” properties.

My statement wasn't very precise. I was more wondering whether we need to unlock them at all. Namely, isn't it at least somewhat expected a property written to by `__clone` shouldn't also be overwritten again? It wasn't clearly specified how this would be handled in the RFC from what I can see. Regardless, I understand why you're doing it so probably just disregard.

> This condition is already there, or did I misunderstand?

No, I just missed it. Apologies.

Reply:
> Namely, isn't it at least somewhat expected a property written to by `__clone` shouldn't also be overwritten again?

The RFC initially proposed that, but this was changed during the discussion (as the RFC authors we did not have strong opinions there):

https://externals.io/message/127353#127394

Reply:
Okay, thanks for the context. Let's resolve this then.